src/middlewared/middlewared/plugins/kubernetes_linux/lifecycle.py

```diff
@@ -0,0 +1,148 @@
+
+        k8s_datasets = set(await self.kubernetes_datasets(config['dataset']))
+        diff = {
+            d['id'] for d in await self.middleware.call('pool.dataset.query', [['id', 'in', list(k8s_datasets)]])
+        } ^ k8s_datasets
+        if diff:
+            raise CallError(f'Missing "{", ".join(diff)}" dataset(s) required for starting kubernetes.')
+
+        locked_datasets = [
+            d['id'] for d in await self.middleware.call('zfs.dataset.locked_datasets')
+            if d['mountpoint'].startswith(f'{config["dataset"]}/') or d['mountpoint'] == config['dataset']
+        ]
+        if locked_datasets:
+            raise CallError(
+                f'Please unlock following dataset(s) before starting kubernetes: {", ".join(locked_datasets)}'
+            )
+        iface_errors = await self.middleware.call('kubernetes.validate_interfaces', config)
+        if iface_errors:
+            raise CallError(f'Unable to lookup configured interfaces: {", ".join([v[1] for v in iface_errors])}')
+
+    @private
+    def status_change(self):
+        config = self.middleware.call_sync('kubernetes.config')
+        if self.middleware.call_sync('service.started', 'kubernetes'):
+            self.middleware.call_sync('service.stop', 'kubernetes')
+
+        if not config['pool']:
+            return
+
+        config_path = os.path.join('/mnt', config['dataset'], 'config.json')
+        clean_start = True
+        if os.path.exists(config_path):
+            with open(config_path, 'r') as f:
+                on_disk_config = json.loads(f.read())
+            clean_start = not all(
+                config[k] == on_disk_config.get(k) for k in ('cluster_cidr', 'service_cidr', 'cluster_dns_ip')
+            )
+
+        if clean_start and self.middleware.call_sync('pool.dataset.query', [['id', '=', config['dataset']]]):
+            self.middleware.call_sync('zfs.dataset.delete', config['dataset'], {'force': True, 'recursive': True})
+
+        self.middleware.call_sync('kubernetes.setup_pool')
+        try:
+            self.middleware.call_sync('kubernetes.status_change_internal')
+        except Exception as e:
+            self.middleware.call_sync('alert.oneshot_create', 'ApplicationsConfigurationFailed', {'error': str(e)})
+            raise
+        else:
+            with open(config_path, 'w') as f:
+                f.write(json.dumps(config))
+
+            self.middleware.call_sync('alert.oneshot_delete', 'ApplicationsConfigurationFailed', None)
+
+    @private
+    async def status_change_internal(self):
+        await self.validate_k8s_fs_setup()
+        await self.middleware.call('service.start', 'docker')
+        # This is necessary because docker daemon requires a couple of seconds after starting to initialise itself
+        # properly, if we try to load images without the delay that will fail and will only correct after a restart
+        await asyncio.sleep(5)
```

Comment:
We have TrueNAS installed on machines slow as hell, with rootfs on USB sticks, shouldn't we use another method to wait for docker to appear?

Reply:
You could just do a call to the docker API and wait while that call fails.